executor/write_test.go

```diff
@@ -2214,3 +2215,79 @@ func (s *testSuite) TestDeferConstraintCheckForInsert(c *C) {
+	defer func() {
+		config.GetGlobalConfig().EnableBatchDML = originConfig
+		atomic.StoreUint64(&kv.TxnEntryCountLimit, originLimit)
+	}()
+	atomic.StoreUint64(&kv.TxnEntryCountLimit, 2)
+	tk.MustExec("set @@session.tidb_dml_batch_size=1;")
+
+	// Test auto commit transaction.
+	config.GetGlobalConfig().EnableBatchDML = true
+	tk.MustExec("insert into t values (1, 'a'), (2, 'b'), (3, 'c')")
+	tk.MustQuery("select * from t order by i").Check(testkit.Rows("1 a", "2 b", "3 c"))
+	tk.MustExec("replace into t values (1, '-a'), (2, '-b'), (3, '-c')")
+	tk.MustQuery("select * from t order by i").Check(testkit.Rows("1 -a", "2 -b", "3 -c"))
+	tk.MustExec("update t set i = -i")
+	tk.MustQuery("select * from t order by i desc").Check(testkit.Rows("-1 -a", "-2 -b", "-3 -c"))
+	tk.MustExec("delete from t")
+	tk.MustQuery("select * from t").Check(testkit.Rows())
+
+	// Test transaction block.
+	tk.MustExec("begin")
+	tk.MustExec("insert into t values (1, 'a'), (2, 'b'), (3, 'c')")
+	tk.MustQuery("select * from t order by i").Check(testkit.Rows("1 a", "2 b", "3 c"))
+	tk.MustExec("replace into t values (1, '-a'), (2, '-b'), (3, '-c')")
+	tk.MustQuery("select * from t order by i").Check(testkit.Rows("1 -a", "2 -b", "3 -c"))
+	tk.MustExec("update t set i = -i")
+	tk.MustQuery("select * from t order by i desc").Check(testkit.Rows("-1 -a", "-2 -b", "-3 -c"))
+	tk.MustExec("delete from t")
+	tk.MustQuery("select * from t").Check(testkit.Rows())
+	tk.MustExec("commit")
+
+	// Test disable batch DML.
+	config.GetGlobalConfig().EnableBatchDML = false
+	_, err := tk.Exec("insert into t values (1, 'a'), (2, 'b'), (3, 'c')")
+	c.Assert(kv.ErrTxnTooLarge.Equal(err), IsTrue)
+	tk.MustQuery("select * from t").Check(testkit.Rows())
+	_, err = tk.Exec("replace into t values (1, '-a'), (2, '-b'), (3, '-c')")
+	c.Assert(kv.ErrTxnTooLarge.Equal(err), IsTrue)
+	tk.MustQuery("select * from t").Check(testkit.Rows())
+	tk.MustExec("insert into t values (1, 'a')")
+	tk.MustExec("insert into t values (2, 'b')")
+	tk.MustExec("insert into t values (3, 'c')")
+	_, err = tk.Exec("update t set i = -i")
+	c.Assert(kv.ErrTxnTooLarge.Equal(err), IsTrue)
+	tk.MustQuery("select * from t order by i").Check(testkit.Rows("1 a", "2 b", "3 c"))
+	_, err = tk.Exec("delete from t")
+	c.Assert(kv.ErrTxnTooLarge.Equal(err), IsTrue)
+	tk.MustQuery("select * from t order by i").Check(testkit.Rows("1 a", "2 b", "3 c"))
+
+	// Test batched retry
+	config.GetGlobalConfig().EnableBatchDML = true
+	atomic.StoreUint64(&kv.TxnEntryCountLimit, 10)
+	tk1 := testkit.NewTestKit(c, s.store)
+	tk1.MustExec("use test")
+	tk.MustExec("set @@session.tidb_dml_batch_size=2;")
+
+	tk.MustExec("begin")
+	tk1.MustExec("update t set j = 'e'")
+	_, err = tk.Exec("update t set j = 'd'")
+	c.Assert(executor.ErrBatchDMLFail.Equal(err), IsTrue, Commentf("error %s", err))
+	tk.MustQuery("select * from t order by i").Check(testkit.Rows("1 e", "2 e", "3 e"))
+
+	// Test normal retry
```

Comment:
"Test normal retry" makes me confused, is there a retry in the following code?

Reply:
tk will retry at line 2291